[로또 2단계] 뭉치 미션 제출합니다.

m6z1 opened the pull request:
안녕하세요! 말리빈 🙌
2단계 미션도 잘 부탁드립니다 ㅎ_ㅎ

# 셀프 체크리스트

<!--
모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
-->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

m6z1 committed:
fix: 로또 생성 로직 수정

m6z1 committed:
style: 함수 반환 값 명시

m6z1 committed:
docs: 로또(수동) 기능 요구 사항 목록 업데이트

m6z1 committed:
remove: 로또 금액 단위 요구 사항 제거

m6z1 committed:
feat: 수동으로 구매할 로또 개수 입력 구현

m6z1 committed:
feat: 수동 로또 번호 입력 로직 구현

m6z1 committed:
feat: 수동/자동 로또 구매 개수 출력

m6z1 committed:
feat: 로또 개수 객체 생성

m6z1 committed:
refactor: 로또 개수 원시 타입을 LottoCount 객체 타입으로 변환

m6z1 committed:
refactor: 로또 생성 객체로 생성 로직 분리

m6z1 committed:
refactor: controller에서 lottos 로 자동/수동 로또들 넘기도록 로직 수정

m6z1 committed:
refactor: try-catch 문을 runCatching 으로 수정

m6z1 committed:
feat: 로또 최소 개수 입력 예외 조건 추가

m6z1 committed:
fix: 로또 최소 개수 1에서 0으로 수정

m6z1 committed:
fix: 수동 로또 번호 예외 시 문제 생긴 로또만 다시 입력받을 수 있도록 로직 수정

m6z1 committed:
style: 변수 타입 명시

m6z1 committed:
fix: 총 상금액 오버 플로우 해결

m6z1 committed:
fix: 수익률 버림 로직으로 수정

m6z1 committed:
refactor: 로또 컨트롤러 내 함수 분리

m6z1 committed:
style: 함수 반환 타입 명시

m6z1 committed:
refactor: 각 로또에서 해당 로또의 등수를 판단하도록 로직 수정

m6z1 committed:
test: 로또 개수 객체 테스트 수정

- 최소 로또 개수가 1에서 0으로 변경됨에 따라 테스트 또한 수정

m6z1 committed:
refactor: 로또 생성(수동/자동) 로직을 하나로 사용

m6z1 committed:
refactor: lotto 객체 data class로 수정

m6z1 commented on src/main/kotlin/lotto/model/LottoMachine.kt:
```diff
@@ -0,0 +1,5 @@
+package lotto.model
+
+class LottoMachine {
```
[#105 comment](https://github.com/woowacourse/kotlin-lotto/pull/105#discussion_r1970207319)

`LottoMachine` 객체를 만들어서 로또를 생성하도록 로직을 분리해보았습니다. 
개발자가 완벽히 제어할 수 있는 환경에서 테스트를 만들도록 lottoNumbers를 매개변수로 받아와서 로또를 생성하도록 수정해보았습니다.

malibinYun replied:
이 객체는 하는 일이 전혀 없어요.
LottoMachine.createLotto() 를 호출하는 것과 Lotto() 생성자를 호출하는 것은 무엇이 다른가요?
컨트롤러에 있는 로또 생성에 대한 일련의 로직을 분리해보세요.

malibinYun replied:
https://github.com/woowacourse/kotlin-lotto/pull/105#discussion_r1964819192
이 객체가 LottoNumbersGenerator 를 가지도록 만들어보세요.
테스트 가능하도록 구조를 변경하면서요.

m6z1 replied:
로또 번호 생성기를 추상화하여, 자동 생성기와 수동 생성기를 구현했습니다. 로또 기계에서 로또를 발행할 때는 로또 번호 생성기와 발행할 로또의 개수를 매개변수로 받아 로또 생성 로직을 `LottoMachine` 객체로 분리하였습니다.

자동 로또 생성에 대한 테스트는 어떻게 해야할까 고민한 결과는 다음과 같습니다.

자동 번호 생성기는 1~45 사이에서 6개의 숫자를 랜덤하게 뽑아 반환하는 방식으로 구현되어 있기에, 테스트에서는 테스트용 로또 번호 생성기를 별도로 구현하여, 자동 로또 발행 테스트 시 해당 생성기가 `1, 2, 3, 4, 5, 6`과 같은 고정된 번호를 반환하도록 하여 테스트를 구현했습니다. 

이전 리뷰 중
> 정확하게 개발자가 제어할 수 있는 환경에서, 어떤 티켓을 발급했는지 일대일 비교할 수 있도록 테스트를 구성해보세요!

라는 리뷰의 접근 방식에 적절한 지 궁금합니다!

- [4171257](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/4171257c692e525bffcd0a0055a5abfe9011ff9f)

malibinYun replied:
로또 머신 자체가 수동 자동 발급을 모두 해주게끔 구성해보는 건 어떨까요?

LottoNumbersGenerator를 상속해서 각 구현체를 만들어 볼 수도 있겠지만, 수동은 항상 객체를 만들어줘야한다는 점이 있지요.
랜덤의 요소만 생성자에서 주입받을 수 있게끔 로또 머신 객체를 구성해보시면 좋겠어요 :)

파라미터로 구성한 것도 하나의 방법이지만, 생성자로도 변경해보시면 좋겠어요!
어떤 차이점이 있을지, 장단점은 무엇일지 고민해보시면서요.

m6z1 committed:
refactor: lottos의 size 필드로 수정

m6z1 commented on src/main/kotlin/lotto/model/Lottos.kt:
```diff
@@ -1,15 +1,18 @@
 ) {
+    val lottoBundle = manualLottos + autoLotts
+
     val size: Int
```
[#105 comment](https://github.com/woowacourse/kotlin-lotto/pull/105#discussion_r1970210005)

전 리뷰에서 달렸던 코멘트에 대한 답변을 작성하고자 합니다.

size에 get()을 두게 되면 해당 `size` 변수는 실제로 필드가 생성되지 않고, lottoBundle.size를 get하는 함수의 역할을 하게 됩니다.
하지만, Lottos는 한 번 생성된 후에, lottoBundle의 사이즈가 변할 일이 없으니 size 를 호출 시마다 lottoBundle.size 를 계산하게 되니 get()을 지우고 필드를 두는 방식으로 수정했습니다!

m6z1 commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -1,6 +1,6 @@
 
-class Lotto(
+data class Lotto(
     val numbers: List<LottoNumber>,
```
[#105 comment](https://github.com/woowacourse/kotlin-lotto/pull/105#discussion_r1970041641) 

해당 코멘트에 대한 답변을 이어서 작성하겠습니다!

Lotto 라는 객체를 사용하는 것은 사용자가 아닌 개발자가 맞지만,
현재 로또 콘솔 게임을 앱으로 옮긴다고 가정하여, 사용자가 중복된 값 (1,2,3,4,5,5) 를 입력했을 때
`입력한 로또 번호 개수는 5개 입니다. 로또 번호는 6개입니다.` 라는 메시지를 보고 중복된 값이 제거되어 입력한 로또 번호 개수가 5개라는 것을 사용자가 판단하기엔 불친절하다고 생각했습니다..! 사용자가 해야 할 행동과 문제점을 명확히 전달해주고자 로또 내부적으로 중복 체크를 로직을 수행하도록 했습니다.

malibinYun replied:
여전히 도메인을 개발하는 사람과 어플리케이션을 사용하는 사람과의 관계가 공감이 가지 않네요.

뷰에서 중복된 값 자체를 받지 못하도록 막는 기능이 생긴다면 여기의 기능은 의미가 있을까요?
Exception 메시지가 정말 애플리케이션 사용자에게 노출되는 메시지인가요?

이 도메인에서 맡아야할 역할일지 깊게 고민해보시면 좋겠네요.
사용성을 좋게 만드는 것과 도메인 로직과는 별개의 것은 아닐지 말이에요.

오히려 list를 통해 Lotto 객체를 생성하고 싶은 개발자를 위해 List의 생성자를 열어주고, 그곳에서 개발자에게 중복을 알려주는 것이라면 납득했을텐데, 도메인 객체에서 왜 애플리케이션 사용자가 언급되는 것인지 공감이 잘 되지 않아요.

m6z1 replied:
제가 핀트를 잘못 짚고 있었던 것 같습니다.

이전 리뷰에서도 말씀해주셨던 것처럼, 도메인의 검증 메시지는 개발자를 위한 것이지, 애플리케이션 사용자를 위한 것이 아닙니다. 그런데 저는 콘솔창의 에러 메시지가 사용자에게 보여지는 것이라고 착각하고 있었습니다.

이번 리뷰를 통해 도메인 객체를 사용하는 것은 개발자라는 점을 확실히 깨달았고, 이전에 언급해주셨던

> Glide 라이브러리 중 Glide.with() 함수에서 여러 옵션을 제공하는 이유가 무엇일까요?

라는 말씀과 연결 지어 생각해보니, 로또 객체 역시 개발자가 다양한 방식으로 생성할 수 있도록 열어두는 것이 맞겠다는 결론을 내렸습니다.

가변인자와 List를 통해 여러 옵션을 두도록 했으며, 해당 옵션들의 로또 번호 중복 검증을 위해 init 에서 중복 검사를 진행하도록 수정해보았습니다.

이 내용에 대한 리뷰가 길어졌는데 이번 기회에 확실히 도메인 객체의 사용처와 역할을 더 명확히 이해할 수 있었습니다. 감사합니다!

-  [2bb20e1](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/2bb20e1853da2a728ab851ae52a89c3510dfa87f)

malibinYun replied:
오 그럼에도 불구하고 주 생성자는 여전히 Set이 아닌 List로 둔 이유가 있으신가요 ?_?

m6z1 reviewed (commented):
안녕하세요 말리빈!
1단계에서 달아주신 리뷰에 대해서 추가로 궁금한 점 및 반영한 부분에 대해 코멘트를 달았습니다!

2단계도 잘 부탁드립니다 !! 🙇‍♀️

malibinYun commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -10,12 +10,6 @@ class WinningLotto(
-        val matchCount = lotto.getMatchCount(winningNumbers)
-        val isMatchedBonus = lotto.isMatchedBonusNumber(bonusNumber)
-        return Rank.findRank(matchCount, isMatchedBonus)
-    }
```
이 메서드의 위치는 어떤 이유로 옮겨진걸까요?_?

m6z1 replied:
로또의 등수를 가지는 것은 로또가 직접 가져야 한다고 생각해서 당첨 번호에서 로또의 등수를 판단하던 로직을 로또 내부로 옮겼었습니다.

그렇지만, lotto 내부로 옮긴 후의 로직을 보니 lotto 에서 winningLotto 의 값을 직접 꺼내 사용하게 되는 일이 생겼습니다. WinningLotto 에게 직접 메시지(몇 개 맞는지 판단해줘/보너스 번호 가지고 있는지 판단해줘)를 던지게 되어도 Lotto 내부의 `getMatchCount`와 `containsNumber`의 메서드와 같은 로직을 사용하기에, 다시 WinningLotto에서 Lotto를 넘겨 주었을 때 `내 (당첨 로또) 기준 너는 몇 등이다.`  라고 등수를 반환하도록 수정했으며, Lotto 내부의 메서드를 재사용하도록 수정해보았습니다!

- [fb6339e](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/fb6339eec736910675109a6c07c57acca199da98)

malibinYun commented on src/main/kotlin/lotto/model/LottoCount.kt:
```diff
@@ -0,0 +1,21 @@
+
+    fun validateLottoMaxCount(totalCount: LottoCount) {
+        require(count <= totalCount.count) { ERROR_MAX_LOTTO_COUNT.format(totalCount.count, count) }
+    }
```
이 메서드는 어떤 역할을 수행하는 건지 알기가 너무 어렵네요.
왜 이 객체 안에 존재해야하는 것인지, 이름으로 어떤 일을 하는 것인지, 왜 자기 자신과 같은 인스턴스를 받아야하는지 등등에서요.

정말 필요한 기능이었나요?
또, Exception을 던져서만 로직을 구성할 수 있는 것이었나요?

m6z1 replied:
오류가 생기면 무조건 Exception 을 던진다고만 생각해서 구현한 로직이었습니다.
말씀해주신 것처럼 객체 안에 존재해야 하는 이유와, 이름으로 어떤 역할을 하는 지 판단하기가 어려워 해당 함수는 현재 제거했습니다.

대신 구매가 가능한지를 판단하는 `isAvailablePurchase` 함수를 구현하여, 구매가 불가능할 경우 사용자에게 다시 입력 받도록 수정해보았습니다.

- [d5c3460](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/d5c34603af40a0d953e2450764b69d234a2dffde)

malibinYun commented on src/main/kotlin/lotto/model/Lotto.kt:
```diff
@@ -26,20 +26,25 @@ class Lotto(
         }
 
-    fun isMatchedBonusNumber(bonusNumber: LottoNumber) = numbers.contains(bonusNumber)
+    private fun isMatchedBonusNumber(bonusNumber: LottoNumber): Boolean = numbers.contains(bonusNumber)
```
이 메서드의 내용은 꼭 "보너스 번호를 포함"하는지 여부만을 알려주는 기능일까요?

m6z1 replied:
로또 번호 중에서 매개변수로 받은 번호가 포함되어 있는지만 판단하는 로직으로
함수의 네이밍 및 매개변수의 네이밍을 number로 수정했습니다.
보너스 번호가 아닌 다른 번호로도 포함되어 있는지 재사용 할 수 있도록 수정했습니다!

- [5fccdba](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/5fccdba500302bf51fd94bc723b86a1451caad71)

malibinYun commented on src/main/kotlin/lotto/model/Lotto.kt:
https://github.com/woowacourse/kotlin-lotto/pull/105#discussion_r1966696689
https://github.com/woowacourse/kotlin-lotto/pull/105#discussion_r1970211550
여긴 이 피드백 까지 4번 코멘트를 남기네요.

아래 isMatchedBonusNumber 메서드에 남긴 코멘트와 함께 연관지어서 고민해보세요.

m6z1 replied:
Lotto 에서 직접 값을 꺼내쓰지 않고, 함수를 재사용 할 수 있도록 수정해보았습니다.
Winning Lotto 의 검증 로직 또한 Lotto의 함수를 재사용 할 수 있도록 수정해보았습니다.

m6z1 replied:
- [0a04d77](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/0a04d774d05e4fde5baf6b6040ad60ac633d7801)
- [79f5da7](https://github.com/woowacourse/kotlin-lotto/pull/151/commits/79f5da79ecb815cb52015b4f736e779d2db2ebfe)

malibinYun reviewed (changes requested):
2단계 미션 고생 많으셨습니다.
여러 코멘트 남겨두었으니 깊게 고민해보세요.
피드백 반영 후 다시 리뷰 요청 부탁드려요.

malibinYun commented:
이전 단계에서 남긴 피드백 질문들에 답이 남겨져 있지 않은 곳들이 있어 아쉬움이 많이 남아요.
이전 피드백들도 아직까지 유효한 것들이 많으니 미션 코드 전체를 종합적으로 고려하시면 좋겠어요.

여러번 남긴 피드백들 중 제대로 나아가지 못하는 것들이 많아요. 시간을 길게 잡고 고민을 많이 해보시면 좋겠어요.
정말 머리가 터질때 까지 고민해보고 모르시겠다면 어떤 고민들이 있었고, 어디서 막힌 것인지 코멘트로 남겨주세요.
뭉치가 어떤 생각을 하고 있는지 알 길이 없기에, 같은 피드백을 남길 수 밖에 없어요.

저는 스스로 답을 찾아내게 만들고 싶어요.
제가 그저 이렇게 하라고 짚어드린다면 매우 쉽게 코드에 반영하겠지만, 쉽게 고친 만큼 쉽게 잊혀져 다시 같은 실수를 반복할 것이기 때문이에요.

m6z1 committed:
refactor: 로또 객체를 내부에서만 생성하도록 로직 수정

m6z1 committed:
refactor: 로또 번호에 해당 번호가 포함되어있는지 판단해주는 함수 네이밍 수정

m6z1 committed:
refactor: 로또 검증 순서 변경

- 변경 전: 로또 번호 중복 검사 > 로또 번호 개수 검사
- 변경 후: 로또 번호 개수 검사 > 로또 번호 중복

m6z1 committed:
refactor: 로또 생성 로직 수정

- LottoMachine.kt: 해당 객체에서 로또 번호 생성기와 로또 개수를 받아 로또를 발행
- 로또 번호 생성기를 추상화하여 수동일 때와 자동일 때 로직을 따로 구현

m6z1 committed:
refactor: 로또 등수 로직을 Lottos에서 처리

m6z1 committed:
refactor: 등수 반환 함수 네이밍 수정

m6z1 committed:
refactor: 로또 객체 생성자로 생성하도록 로직 수정

m6z1 committed:
refactor: 다른 로또 개수와 일치하는 여부 함수의 매개변수 네이밍 수정

m6z1 committed:
refactor: 당첨 로또 객체의 당첨 번호 검증 로직을 로또에서 진행

m6z1 committed:
refactor: 로또 등수 반환 로직 당첨 번호로 이동

m6z1 committed:
feat: 중복된 로또 번호 에러 메시지에 추가

m6z1 committed:
fix: 수동 로또 번호 검증 후 재입력 로직 수정

m6z1 committed:
refactor: 로또 개수 검증 로직 수정

m6z1 committed:
refactor: 로또 일치 개수 판단 로직 수정

- containsNumber 를 재사용하여 로직을 수행

m6z1 committed:
test: 로또 내부 로직 테스트 추가

- 로또 일치 개수 반환 테스트
- 로또 번호 포함 여부 테스트

m6z1 committed:
test: 당첨 로또 등수 반환 테스트 추가

m6z1 commented on src/main/kotlin/lotto/controller/LottoController.kt:
```diff
@@ -1,50 +1,111 @@
+        return lottoMachine.createLottos(manualLottoCount, numbersGenerator)
+    }
+
+    private fun getManualNumbersBundle(lottoCount: LottoCount): List<List<LottoNumber>> = List(lottoCount.count) { getValidLottoNumbers() }
+
+    private fun getValidLottoNumbers(): List<LottoNumber> =
+        runCatching {
+            val numbers: List<LottoNumber> = inputView.readLottoNumbers()
+            Lotto(numbers)
+            return numbers
```
현재, 수동 및 자동 로또 번호 생성기를 추상화하여 구현하는 과정에서 몇 가지 고민이 생겼습니다.

 **1. 현재 구조**

- 수동 로또는 사용자가 모든 번호를 입력한 후, `수동 로또 번호 생성기`를 통해 번호를 하나씩 가져와 `LottoMachine`에서 `Lotto`를 일괄적으로 생성합니다.
- 그러나 이 방식에서는 모든 번호 입력 후 `LottoMachine`에서 `Lotto`를 생성하기 때문에, 사용자가 입력할 때마다 즉시 검증되지 않습니다.
- 이를 해결하기 위해 현재는 수동 로또 번호 입력 시점에서 `Lotto(numbers)`를 호출하여 즉시 검증하고 있습니다.

**2. 고민되는 점**

- `Lotto(numbers)`를 이용해 검증을 진행하는데, 이 과정에서 불필요한 `Lotto` 인스턴스가 여러 번 생성되고 있습니다.
- 그렇다면 입력할 때마다 `LottoMachine.createLotto(lottoNumbers)`를 호출하여 로또 하나 씩 생성해야 할까? 라는 생각이 들었습니다.
- 하지만, 위 로직으로 구현하게 되면 결국 `LottoMachine`이 필요할까? `LottoMachine.createLotto()`를 호출하는 것과 `Lotto(numbers)`를 직접 생성하는 것은 단순한 포장 차이로 굳이 LottoMachine 이 필요하지 않을 것 같다는 생각이 들었습니다.

- 만약 `LottoMachine` 없이 Controller에서 직접 `Lotto`를 생성하면, Controller의 역할이 너무 커진다고 생각했습니다. 

고민의 굴레에 끝에 결국엔 처음 구현한대로 리뷰 요청을 드립니다.

어떤 구조로 로또를 생성하는 것이 더 적절한지, 현재 구조에 대한 방향성이 맞는 방향인지 말리빈의 의견이 궁금합니다.. ㅠㅠ

malibinYun replied:
오호~ 저는 지금의 고민이 굉장히 유의미한 고민이라 생각해요.
어디든 장단점과 트레이드 오프가 생길 수 밖에 없다는 점을 이 고민으로 완전히 체험하셨다는 생각이 들어요.
머리가 부서지도록 고민해서 자신만의 기준이 만들어지길 기대해요 :)

저는 로또 머신이라는 객체가 존재함으로써 (로또의 생성을 담당하는 객체)
로또를 만들때는 항상 로또의 생성자를 직접 불러야하나? 아니면 다른 객체가 있나? 라는 고민을 아예 배제시켜줄 수 있다 생각해요.
생성해주는 건 생성기에 다 있겠지 라는 생각을 하게끔 유도하는 것이지요.

뭉치의 생각은 어떨지 궁금하네요

m6z1 committed:
style: 변수 타입 명시

m6z1 reviewed (commented):
안녕하세요, 말리빈!
귀한 리뷰 감사합니다. 🙇‍♀️

1단계 미션에서 받은 피드백을 이번 2단계 피드백을 받고서야 이해한 것 같습니다. 이전에는 피드백의 의미를 완벽하게 파악하지 못해 피드백을 잘 반영하지 못한 것 같습니다. 이번 2단계 피드백 이후에 더 곰곰이 생각해 본 후 코드의 개선 방향이나 구조를 수정해보았습니다. 

궁금한 점과 고민의 과정을 코드 코멘트로 달았습니다.

또한, 이번 미션이 TDD로 진행하기 였음에도 리팩토링 과정은 본 코드를 수정 후 테스트 코드를 리팩토링 하게 되었습니다.
본 코드 리팩토링과 테스트 코드 커밋이 따로 되어있는 부분이 있습니다. 다음 TDD 시에는 리팩토링 또한 TDD로 진행하여 테스트와 같이 커밋할 수 있도록 하겠습니다😢

malibinYun commented on src/main/kotlin/lotto/model/WinningLotto.kt:
```diff
@@ -6,7 +6,7 @@ class WinningLotto(
 ) {
     init {
         require(
-            winningNumbers.numbers.contains(bonusNumber).not(),
+            winningNumbers.containsNumber(bonusNumber).not(),
```
이 반복 학습으로 객체에게 메시지를 던지는 것에 익숙해지셨으면 좋겠네요 👍

malibinYun reviewed (approved):
로또 미션 완주하시느라 고생 많으셨어요.
다음 블랙잭 미션은 매우 어려우니 이만 마무리하도록 하겠습니다.
궁금한 점이 있으시다면 언제든지 DM 주셔도 좋으니 메시지 남겨주세요.
미션 진행하시면서 고민했던 내용들이 앞으로도 큰 도움이 되었으면 좋겠네요 :)
